storage/client/client.go

```diff
@@ -529,25 +577,55 @@ func (c *StorageClient) Account(ctx context.Context, r *AccountRequest) (*Accoun
 	a := Account{
 		Allowances: []Allowance{},
 	}
+	var availableStr string
+	var escrowStr string
+	var debondingStr string
+	var delegationsBalanceStr string
+	var debondingDelegationsBalanceStr string
 	if err := c.db.QueryRow(
 		ctx,
 		qf.AccountQuery(),
 		r.Address.String(),
 	).Scan(
 		&a.Address,
 		&a.Nonce,
-		&a.Available,
-		&a.Escrow,
-		&a.Debonding,
-		&a.DelegationsBalance,
-		&a.DebondingDelegationsBalance,
+		&availableStr,
+		&escrowStr,
+		&debondingStr,
+		&delegationsBalanceStr,
+		&debondingDelegationsBalanceStr,
 	); err != nil {
 		c.logger.Info("row scan failed",
 			"request_id", ctx.Value(RequestIDContextKey),
 			"err", err.Error(),
 		)
 		return nil, common.ErrStorageError
 	}
+	available, err := c.deserializeToBigIntOrLog(ctx, availableStr)
+	if err != nil {
+		return nil, common.ErrStorageError
+	}
+	escrow, err := c.deserializeToBigIntOrLog(ctx, escrowStr)
+	if err != nil {
+		return nil, common.ErrStorageError
+	}
+	debonding, err := c.deserializeToBigIntOrLog(ctx, debondingStr)
+	if err != nil {
+		return nil, common.ErrStorageError
+	}
+	delegationsBalance, err := c.deserializeToBigIntOrLog(ctx, delegationsBalanceStr)
+	if err != nil {
+		return nil, common.ErrStorageError
+	}
+	debondingDelegationsBalance, err := c.deserializeToBigIntOrLog(ctx, debondingDelegationsBalanceStr)
+	if err != nil {
+		return nil, common.ErrStorageError
+	}
+	a.Available = *available
+	a.Escrow = *escrow
+	a.Debonding = *debonding
+	a.DelegationsBalance = *delegationsBalance
+	a.DebondingDelegationsBalance = *debondingDelegationsBalance
```

Comment:
Alternatively, 

```
fields := make([]*big.Int, 0)
	for _, s := range []string{availableStr, escrowStr, debondingStr, delegationsBalanceStr, debondingDelegationsBalanceStr} {
		num, err := c.deserializeToBigIntOrLog(ctx, s)
		if err != nil {
			return nil, common.ErrStorageError
		}
		fields = append(fields, num)
	}
	a.Available = *fields[0]
	a.Escrow = *fields[1]
	a.Debonding = *fields[2]
	a.DelegationsBalance = *fields[3]
	a.DebondingDelegationsBalance = *fields[4]
```

but this felt less clean

Reply:
The preferred way would be to use pgx's support for custom types. I haven't used it before, but it should go like this:

```go
type DbBigInt big.Int  // wrap the big.Int type so we can add methods to it

// Implements pgtype.TextDecoder
func (dst *DbBigInt) DecodeText(_ci *pgtype.ConnInfo, src []byte) error {
	*dst, err := i.SetString(string(src), 10)
	return err
}

var available DbBigInt
db.[...].Scan(..., &available)
a.available = available.(big.Int)
```

Caveats apply.
- The code probably has issues; this is written from looking at docs, I haven't tried to compile anything.
- From looking at the source, the `DecodeText` approach only works if postgres exposes the field as text (there's also `DecodeBinary`), and I don't know what postgres does for our types.
- I don't know how this handles NULLs; the input to `SetString` is `""` in that case.

There is already a built-in [`Numeric`](https://github.com/jackc/pgx/blob/v3.6.2/pgtype/numeric.go#L46) (meant to be a go representation of the postgres NUMERIC), and we could potentially use that instead of a custom type, but for some reason its `toBigInt` method is private :confused:. Also, we have custom types in postgres (like UINT63) that do wrap NUMERIC, but I think they're still special enough that pgx wouldn't know to map them to `pgtype.Numeric`.

It would be cleanest if we were using a similar `EncodeText()` approach in the other direction (writing bigints into the DB), but in that direction, it doesn't make the code cleaner really. Right now, we call `.String()` on those big.Ints, and the "proper" way would be to convert them to some `DbBigInt`, which would then convert it to a string (or whatever our db engine needs, which could theoretically be something else in a very hypothetical future) behind the scenes. But either way, most callsites that deal with the db would still have to perform a conversion.

Reply:
After some trial and error, it seems like pgx attempts to Scan our custom db type (`UINT_NUMERIC`) using the [BinaryDecoder interface](https://pkg.go.dev/github.com/jackc/pgtype@v1.13.0#BinaryDecoder), which means we have to implement `DecodeBinary`, eg in 

```
type DbBigInt struct {
	inner *big.Int
}

func (dst *DbBigInt) DecodeBinary(_ci *pgtype.ConnInfo, src []byte) error {
	if src == nil {
		dst.inner = nil
		return nil
	}

	dst.inner = big.NewInt(0)
	dst.inner.SetBytes(src) // This doesn't actually work
	return nil
}
```

Turns out the byte array postgres presents to pgx is the direct representation of our custom `UINT_NUMERIC` -> `NUMERIC(1000,0)`. It's a bit unfortunate that our application needs to know about our custom internal postgres types, but I don't expect us to change the definition of `UINT_NUMERIC` so I'll keep working on this.

Reply:
what was the issue here? this thread is attached to a line that says

```go
	a.DebondingDelegationsBalance = *debondingDelegationsBalance
```

and Andrew says there's a way to do custom decoding of byte arrays but Mitja posted a link showing that somewhere in pgx there's already a routine for decoding numerics-------which we use, don't we?

and yet there was talk of creating custom types to wrap *big.Int and talk of copying pgx code into our codebase?

Reply:
also,

```go
func iiii(s Numeric) *big.Int {
  var r big.Rat
  s.AssignTo(&r)
  assert(r.IsInt()) // numeric(*, 0) must be integer
  return r.Num
}
```

Reply:
> what was the issue here?

Prior to this PR, pgx is `Scanning` postgres db `Numeric`s into golang `uint64`, which is problematic bc of overflow. 

My first attempt to remedy this was to `Scan` each field into a tmp string, and then convert the string to a big.Int. This was somewhat unwieldy because we had to a lot of the error-checking ourselves and it made some functions quite long/repetitive. It also doesn't work because postgres doesn't expose our custom `UINT_NUMERIC` types as strings, it only exposes the raw bytes.

Mitja suggested wrapping `big.Int` in a wrapper class `DbBigInt` and implementing the `pgtype.BinaryDecoder` interface for numerics on that class, which would allow us to leverage pgx's native error checking when `Scan`ning to `DbBigInt`. This allowed us to skip a lot of the manual error checks, and is probably more idiomatic.

However, implementing `BinaryDecoder` is gnarly to do ourselves since the binary format of postgres Numerics is complex. The golang `pgtype.Numeric` does implement `pgtype.BinaryDecoder`, and ideally we'd just `Scan` postgres values to `pgtype.Numeric` and then convert that to `big.Int` (our internal representation). Unfortunately, the `pgtype.Numeric.toBigInt()` method is private, so we have to do it ourselves. 

One option is to simply copy their private `pgtype.Numeric.toBigInt()` into our codebase. We could also do it the way you suggested above :)

Reply:
@pro-wh about your `iiii()` suggestion: That's how I would expect/want it to work, but `Numeric.AssignTo(dst)` [doesn't support](https://github.com/jackc/pgx/blob/a413de981978e4100d5736af7aaa2b392511612f/pgtype/numeric.go#L118) `big.Int` or `big.Rat` dst types, unless I'm misreading.